libs/platform/ebpf_ring_buffer.h

```diff
@@ -0,0 +1,139 @@
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+    typedef struct _ebpf_ring_buffer_record
+    {
+        struct
+        {
+            uint8_t locked : 1;
+            uint8_t discarded : 1;
+            uint32_t length : 30;
+        } header;
+        uint8_t data[1];
+    } ebpf_ring_buffer_record_t;
+
+    typedef struct _ebpf_ring_buffer ebpf_ring_buffer_t;
+
+    /**
+     * @brief Allocate a ring_buffer with capacity.
+     *
+     * @param[out] ring_buffer Pointer to buffer that holds ring buffer pointer on success.
+     * @param[in] capacity Size in bytes of ring buffer.
+     * @retval EPBF_SUCCESS Successfully allocated ring buffer.
+     * @retval EBPF_NO_MEMORY Unable to allocate ring buffer.
+     */
+    ebpf_result_t
+    ebpf_ring_buffer_create(_Outptr_ ebpf_ring_buffer_t** ring_buffer, size_t capacity);
+
+    /**
+     * @brief Free a ring buffer.
+     *
+     * @param[in] ring_buffer Ring buffer to free.
+     */
+    void
+    ebpf_ring_buffer_destroy(_Frees_ptr_opt_ ebpf_ring_buffer_t* ring_buffer);
+
+    /**
+     * @brief Write out a variable sized record to the ring buffer.
+     *
+     * @param[in,out] ring_buffer Ring buffer to write to.
+     * @param[in] data Data to copy into record.
+     * @param[in] length Length of data to copy.
+     * @retval EPBF_SUCCESS Successfully wrote record ring buffer.
+     * @retval EBPF_INVALID_ARGUMENT Unable to output to ring buffer.
+     */
+    ebpf_result_t
+    ebpf_ring_buffer_output(
+        _Inout_ ebpf_ring_buffer_t* ring_buffer, _In_reads_bytes_(length) uint8_t* data, size_t length);
+
+    /**
+     * @brief Query the current ready and free offsets from the ring buffer.
+     *
+     * @param[in] ring_buffer Ring buffer to query.
+     * @param[out] consumer Offset of the first buffer that can be consumed.
+     * @param[out] producer Offset of the next buffer to be produced.
+     */
+    void
+    ebpf_ring_buffer_query(_In_ const ebpf_ring_buffer_t* ring_buffer, _Out_ size_t* consumer, _Out_ size_t* producer);
+
+    /**
+     * @brief Mark one or more records in the ring buffer as returned to the ring.
+     *
+     * @param[in,out] ring_buffer Ring buffer to update.
+     * @param[in] length Length of bytes to return to the ring buffer.
+     * @retval EPBF_SUCCESS Successfully returned records to the ring buffer.
+     * @retval EBPF_INVALID_ARGUMENT Unable to return records to the ring buffer.
+     */
+    ebpf_result_t
+    ebpf_ring_buffer_return(_Inout_ ebpf_ring_buffer_t* ring_buffer, size_t length);
+
+    /**
+     * @brief Get pointer to the ring buffer shared data.
+     *
+     * @param[in] ring_buffer Ring buffer to query.
+     * @param[out] buffer Pointer to ring buffer data.
+     * @retval EPBF_SUCCESS Successfully mapped the ring buffer.
+     * @retval EBPF_INVALID_ARGUMENT Unable to map the ring buffer.
+     */
+    ebpf_result_t
+    ebpf_ring_buffer_map_buffer(_In_ const ebpf_ring_buffer_t* ring_buffer, _Outptr_ uint8_t** buffer);
```

Comment:
Why is this API required that exposes the shared buffer? Why does ebpf_ring_buffer_next_record take  the shared buffer instead of ring buffer?

Reply:
Do we want the out param type to be an opaque void** ?

Reply:
This is required so that we can achieve zero-copy semantics. Producer (eBPF program) writes directly into the ring buffer (via memory returned by bpf_ring_buf_reserve/commit) and the consumer (user application) reads directly from the ring buffer. 

uint8_t because the value returned is a byte-addressable uint8_t array.